### PR TITLE
Boats refactoring

### PR DESCRIPTION
**Pull request by Nordsoft91**

### CODE REVIEW

**Comment by rilian-la-te** [comment]
@Nordsoft91 will be possible to load boat animation from mod?

**Comment by Nordsoft91** [comment]
Yes

File: lib/GameConstants.h [hunk 1]
```diff
@@ -1241,6 +1241,27 @@ class BattleField : public BaseForID<BattleField, si32>
 	DLL_LINKAGE static BattleField fromString(const std::string & identifier);
 };
 
+class TransportId
```
Comment (rilian-la-te) [inline lib/GameConstants.h:1244]: Why this is ID_LIKE and not Identifier?

File: lib/mapObjects/CGHeroInstance.cpp [hunk 2]
```diff
@@ -935,18 +935,14 @@ si32 CGHeroInstance::getManaNewTurn() const
...
 	{
-	case EAlignment::GOOD:
-		return 1;
-	case EAlignment::EVIL:
-		return 0;
-	case EAlignment::NEUTRAL:
-		return 2;
-	default:
-		throw std::runtime_error("Wrong alignment!");
+		case EAlignment::EVIL : return TransportId::ETransportId::BOAT_EVIL;
```
Comment (IvanSavenko) [inline lib/mapObjects/CGHeroInstance.cpp:942]: Move to hero class config?

File: lib/mapObjects/CGTownInstance.cpp [hunk 3]
```diff
@@ -1085,16 +1085,15 @@ void CGTownInstance::clearArmy() const
...
 
-int CGTownInstance::getBoatType() const
+TransportId CGTownInstance::getTransportType() const
 {
 	switch (town->faction->alignment)
 	{
-	case EAlignment::EVIL : return 0;
-	case EAlignment::GOOD : return 1;
-	case EAlignment::NEUTRAL : return 2;
+		case EAlignment::EVIL : return TransportId::ETransportId::BOAT_EVIL;
```
Comment (IvanSavenko) [inline lib/mapObjects/CGTownInstance.cpp:1092]: Move to config?

File: lib/mapObjects/CObjectHandler.cpp [hunk 4]
```diff
@@ -457,16 +457,16 @@ IBoatGenerator::EGeneratorState IBoatGenerator::shipyardStatus() const
...
 		return GOOD; //OK
-	else if(t->blockingObjects.front()->ID == Obj::BOAT)
+	else if(t->blockingObjects.front()->ID == Obj::TRANSPORT)
 		return BOAT_ALREADY_BUILT; //blocked with boat
 	else
 		return TILE_BLOCKED; //blocked
 }
 
-int IBoatGenerator::getBoatType() const
+TransportId IBoatGenerator::getTransportType() const
```
Comment (IvanSavenko) [inline lib/mapObjects/CObjectHandler.cpp:466]: Perhaps make this method pure virtual and reimplement it in CGShipyard?

All other inheritors already use own implementation anyway.
Also - would be great to be able to define what "transport" is built in shipyard in config.

**Review by IvanSavenko — COMMENTED** [review]
Nice!

- Have other "transport" types (for layers other than water) have been tested?

- Perhaps transport should define movement point amount and not hero? ~Math is clearly different (fixed 2500 for boats vs based on creature speed for land)~ (EDIT: might we wrong, can't remember) and might be useful for other transport types, or even for "special" boats that are faster/slower than standard.

File: client/windows/GUIClasses.cpp [hunk 5]
```diff
@@ -1089,7 +1089,7 @@ CShipyardWindow::CShipyardWindow(const TResources & cost, int state, int boatTyp
 	std::string boatFilenames[3] = {"AB01_", "AB02_", "AB03_"};
 
 	Point waterCenter = Point(bgWater->pos.x+bgWater->pos.w/2, bgWater->pos.y+bgWater->pos.h/2);
-	bgShip = std::make_shared<CAnimImage>(boatFilenames[boatType], 0, 7, 120, 96, 0);
+	bgShip = std::make_shared<CAnimImage>(boatFilenames[boatType.getNum()], 0, 7, 120, 96, 0);
```
[the remark this answers is not in the document]
Comment (Nordsoft91) [reply client/windows/GUIClasses.cpp:1092]: doesn't make much sense - shipyard will be revised to support new transports

File: lib/GameConstants.h (hunk 1 above)
Comment (Nordsoft91) [reply lib/GameConstants.h:1244]: what do you mean?

File: lib/mapObjects/CGHeroInstance.cpp (hunk 2 above)
Comment (Nordsoft91) [reply lib/mapObjects/CGHeroInstance.cpp:942]: out of this PR

File: lib/mapObjects/CGTownInstance.cpp (hunk 3 above)
Comment (Nordsoft91) [reply lib/mapObjects/CGTownInstance.cpp:1092]: out of this PR

File: lib/mapObjects/CObjectHandler.cpp (hunk 4 above)
Comment (Nordsoft91) [reply lib/mapObjects/CObjectHandler.cpp:466]: this change is only about boats, not about shipyards

**Comment by Nordsoft91** [comment]
> Have other "transport" types (for layers other than water) have been tested?

Yes, have a mod prototype for that.

> Perhaps transport should define movement point amount and not hero? Math is clearly different (fixed 2500 for boats vs based on creature speed for land) (EDIT: might we wrong, can't remember) and might be useful for other transport types, or even for "special" boats that are faster/slower than standard.

Right, that's could be more useful, but for now it's solved by adding/removing bonuses, related to movement.
I'll investigate how complex would be to change logic, but not everything at once

File: lib/GameConstants.h (hunk 1 above)
Comment (rilian-la-te) [reply lib/GameConstants.h:1244]: For example - in this file you can find ETerrainId and TerrainId, which done by identifiers.

**Comment by rilian-la-te** [comment]
@IvanSavenko now army movement and all you described works on bonuses, so, it can simply be another node, why not. Like silent artifact.

File: client/windows/GUIClasses.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (IvanSavenko) [reply client/windows/GUIClasses.cpp:1092]: Understood. This shipyard revision is in your plans for 1.3, correct?

File: lib/mapObjects/CGHeroInstance.cpp (hunk 2 above)
Comment (Nordsoft91) [reply lib/mapObjects/CGHeroInstance.cpp:942]: no, didn't plan to refactor hero unless there will be such request

File: config/objects/moddables.json [hunk 6]
```diff
@@ -115,9 +115,38 @@
 			}
 		},
 		"types" : {
-			"evil" : { "index" : 0,	"rmg" : { "mapLimit" : 64 } },
-			"good" : { "index" : 1, "rmg" : { "mapLimit" : 64 } },
-			"neutral" : { "index" : 2, "rmg" : { "mapLimit" : 64 } },
+			"evil" : 
+			{ 
+				"index" : 0,	
+				"rmg" : { "mapLimit" : 64 },
```
Comment (DjWarmonger) [inline config/objects/moddables.json:121]: RMG threats Boat as a special object and doesn't really check this limit. Engine doesn't have such limitations either, afaik.

File: lib/GameConstants.h [hunk 7]
```diff
@@ -756,7 +756,7 @@ class Obj
 		ARTIFACT = 5,
 		PANDORAS_BOX = 6,
 		BLACK_MARKET = 7,
-		BOAT = 8,
+		TRANSPORT = 8,
```
Comment (DjWarmonger) [inline lib/GameConstants.h:759]: But why? Everyone knows what Boat is, but "Transport" is unclear to say at least.

Comment (krs0) [reply lib/GameConstants.h:759]: That one would be Airship or skyship...
So Boat (aka Ship) could be VEHICLE or CRAFT in generic form and Ship in particular cases when you want to refer to water vehicles.

Comment (Nordsoft91) [reply lib/GameConstants.h:759]: Maybe only renaming (without functional changes) could be in separate PR if we agreed on something. It's a lot of stuff to rename, so I'm with @DjWarmonger here

File: lib/GameConstants.h (hunk 1 above)
Comment (Nordsoft91) [reply lib/GameConstants.h]: fixed, thanks!

**Pull request merged**